aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +27,348 @@ namespace at {
-          break; \
-        } \
-          else \
-        { \
-          /* Reset the pointer to the beginning of the chunk defined by this dimension */ \
-          ATENSOR##_data -= ATENSOR##_counter[ATENSOR##_i]*ATENSOR##_strides[ATENSOR##_i]; \
-          ATENSOR##_counter[ATENSOR##_i] = 0; \
-        } \
-      } \
-      else \
-        break; \
-    } \
-    ATENSOR##_i = 0; \
+template <typename T, int N>
+struct AT_API strided_tensor_iter_fixed {
+ public:
+  T* data_ = NULL;
+  int64_t dim_;
+
+  int64_t counter_[N];
+  int64_t sizes_[N];
+  int64_t strides_[N];
+
+  strided_tensor_iter_fixed(strided_tensor_iter_fixed const&) = delete;
+  void operator=(strided_tensor_iter_fixed const& x) = delete;
+  strided_tensor_iter_fixed(strided_tensor_iter_fixed&&) = default;
+  strided_tensor_iter_fixed(Tensor& tensor, bool sort_strides = false)
+      : data_(tensor.data<T>()), dim_(tensor.ndimension()) {
+    memset(counter_, 0, sizeof(int64_t) * N);
+    int64_t max_dim = tensor.ndimension();
+    dim_ = 0;
+    for (int64_t i = 0; i < max_dim; i++) {
+      int64_t size = tensor.size(i);
+      int64_t stride = tensor.stride(i);
+      while (i + 1 < max_dim &&
+             tensor.stride(i) == tensor.size(i + 1) * tensor.stride(i + 1)) {
+        size = size * tensor.size(i + 1);
+        stride = tensor.stride(i + 1);
+        i++;
+      }
+      sizes_[dim_] = size;
+      strides_[dim_] = stride;
+      dim_++;
+    }
   }
+};
 
-template <typename scalar1, typename scalar2, typename Op>
-void CPU_tensor_apply2_dim(Tensor& tensor1, Tensor& tensor2, int64_t dim, Op op) {
-  checkBackend("CPU_tensor_apply2", {tensor1, tensor2}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  if(t1_numel != t2_numel) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and " << tensor2.sizes()
-        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+template <typename T>
+struct AT_API strided_tensor_iter {
+ private:
+ public:
+  T* data_ = NULL;
+  const int64_t dim_;
+
+  std::vector<int64_t> counter_;
+  std::vector<int64_t> sizes_;
+  std::vector<int64_t> strides_;
+
+  strided_tensor_iter(strided_tensor_iter const&) = delete;
+  void operator=(strided_tensor_iter const& x) = delete;
+  strided_tensor_iter(strided_tensor_iter&&) = default;
+  strided_tensor_iter(Tensor& tensor)
+      : data_(tensor.data<T>()),
+        dim_(tensor.ndimension()),
+        counter_(dim_, 0),
+        sizes_(tensor.sizes()),
+        strides_(tensor.strides()) {}
+};
+
+template <typename Arg>
+AT_API inline void forward(int64_t offset, Arg& ut) {
```

Comment:
What's this function for? There's no way to tell what `Arg& ut` is supposed to be. Can you please use a more descriptive name, and add a comment?

Reply:
I'll definitely add a lot more comments overall once all CI tests run clear.